README.md
ASCII art movie Telnet player
=============================

Can stream an ~20 minutes ASCII movie via Telnet emulation
as stand alone server or via xinetd daemon. 

<img src="screenshots/example.gif?raw=true" width=500>

Wanna see it in action? Just watch http://asciinema.org/a/3132


Tested with Python 2.6+, Python 3.5+

Original art work : Simon Jansen [http://www.asciimation.co.nz/](http://www.asciimation.co.nz/)  
Telnetification & Player coding : Martin W. Kirst  
Python3 Update: Ryan Jarvis  

Command line parameters
-----------------------

See program output:

    $ python ascii_telnet_server.py --help
    Usage: ascii_telnet_server.py [options]
    Options:
      -h, --help            show this help message and exit
      --standalone          Run as stand alone multi threaded TCP server (default)
      --stdout              Run with STDIN and STDOUT, for example in XINETD
                            instead of stand alone TCP server. Use with python
                            option '-u' for unbuffered STDIN STDOUT communication
      -f FILE, --file=FILE  Text file containing the ASCII movie
      -i INTERFACE, --interface=INTERFACE
                            Bind to this interface (default '0.0.0.0', all
                            interfaces)
      -p PORT, --port=PORT  Bind to this port (default 23, Telnet)
      -v, --verbose         Verbose (default for TCP server)
      -q, --quiet           Quiet! (default for STDIN STDOUT server)


Run as stand alone server
-------------------------

Simple call this Python script by using the sample movie file:

    $> python ascii_telnet_server.py --standalone -f ../sample_movies/sw1.txt
    Running TCP server on 0.0.0.0:23
    Playing movie sw1.txt
   

Run as xinetd program
---------------------

place this configuration into `/etc/xinetd.d/telnet`:

    # default: on
    # description: An telnet service playing an ASCII movie, Star Wars Episode 4 
    service telnet
    {
            disable         = no
            socket_type     = stream
            protocol        = tcp
            port            = 23
            user            = root
            wait            = no
            instances       = 10
    
            log_type        = FILE /var/log/asciiplayer
            log_on_success  += PID HOST DURATION
            log_on_failure  = HOST
            server          = /usr/bin/python
            server_args     = -u -OO /opt/asciiplayer/ascii_telnet_server.py -f /opt/asciiplayer/sw1.txt --stdout
    }



CHANGES.txt
v1.0, 11/21/2106 -- Initial PyPI release


tests/test_timebar.py
# coding=utf-8
import pytest

from ascii_theater.ascii_telnet.ascii_movie import TimeBar


class TestTimeBar(object):
    def test_empty_timebar(self):
        tb = TimeBar(0, 10)
        assert tb._empty_timebar == "<        >"

    def test_long_timebar(self):
        tb = TimeBar(0, 100)
        assert tb._empty_timebar == "<                                " \
                                    "                                 " \
                                    "                                 >"

    def test_too_short_timebar(self):
        with pytest.raises(ValueError) as excinfo:
            TimeBar(0, 1)
            assert str(excinfo.values) == "This TimeBar is too short for these decorators: < o >"

    def test_short_timebar(self):
        tb = TimeBar(0, 3)
        assert tb._empty_timebar == "< >"

    def test_fancy_decorators(self):
        tb = TimeBar(0, 15, left_decorator=u"*~~*~~", right_decorator=u"Œ")
        assert tb._empty_timebar == u"*~~*~~        Œ"
        assert len(tb._empty_timebar) == 15

    def test_custom_spacer(self):
        tb = TimeBar(0, 15, spacer=".")
        assert tb._empty_timebar == "<.............>"

    def test_get_marker_position_easy(self):
        tb = TimeBar(100, 102)
        assert tb.get_marker_postion(0) == 0
        assert tb.get_marker_postion(1) == 1
        assert tb.get_marker_postion(10) == 10
        assert tb.get_marker_postion(20) == 20
        assert tb.get_marker_postion(30) == 30
        assert tb.get_marker_postion(40) == 40
        assert tb.get_marker_postion(50) == 50
        assert tb.get_marker_postion(99) == 99
        assert tb.get_marker_postion(100) == 100

    def test_get_marker_position_harder(self):
        tb = TimeBar(100, 100)
        assert tb.get_marker_postion(0) == 0
        assert tb.get_marker_postion(1) == 1
        assert tb.get_marker_postion(10) == 10
        assert tb.get_marker_postion(20) == 20
        assert tb.get_marker_postion(30) == 29
        assert tb.get_marker_postion(40) == 39
        assert tb.get_marker_postion(50) == 49
        assert tb.get_marker_postion(99) == 97
        assert tb.get_marker_postion(100) == 98

    def test_get_marker_position_small(self):
        tb = TimeBar(100, 10)
        assert tb.get_marker_postion(0) == 0
        assert tb.get_marker_postion(1) == 0
        assert tb.get_marker_postion(10) == 1
        assert tb.get_marker_postion(20) == 2
        assert tb.get_marker_postion(30) == 2
        assert tb.get_marker_postion(40) == 3
        assert tb.get_marker_postion(50) == 4
        assert tb.get_marker_postion(99) == 8
        assert tb.get_marker_postion(100) == 8

    def test_timebar_with_marker(self):
        tb = TimeBar(100, 102)

        assert len(tb.get_timebar(0)) == 102
        assert tb.get_timebar(0).index("o") == 1
        assert tb.get_timebar(0) == "<o                                                                                                   >"

        assert len(tb.get_timebar(10)) == 102
        assert tb.get_timebar(10).index("o") == 11
        assert tb.get_timebar(10) == "<          o                                                                                         >"

        assert len(tb.get_timebar(99)) == 102
        assert tb.get_timebar(99).index("o") == 100

        assert len(tb.get_timebar(100)) == 102
        assert tb.get_timebar(100).index("o") == 100

    def test_timebar_with_marker_short(self):
        tb = TimeBar(100, 8)

        assert len(tb.get_timebar(0)) == 8
        assert tb.get_timebar(0).index("o") == 1
        assert tb.get_timebar(0) == "<o     >"

        assert len(tb.get_timebar(10)) == 8
        assert tb.get_timebar(10).index("o") == 2
        assert tb.get_timebar(10) == "< o    >"

        assert len(tb.get_timebar(50)) == 8
        assert tb.get_timebar(50).index("o") == 4
        assert tb.get_timebar(50) == "<   o  >"

        assert len(tb.get_timebar(99)) == 8
        assert tb.get_timebar(99).index("o") == 6
        assert tb.get_timebar(99) == "<     o>"

        assert len(tb.get_timebar(100)) == 8
        assert tb.get_timebar(100).index("o") == 6
        assert tb.get_timebar(100) == "<     o>"

    def test_frame_num_larger_than_timebar_frame_count(self):
        tb = TimeBar(100, 102)
        bad_frame_timebar = tb.get_timebar(104)
        assert len(bad_frame_timebar) == 102
        assert bad_frame_timebar == "<                                                                                                   o>"


ascii_theater/ascii_telnet/ascii_server.py
# coding=utf-8
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#  Copyright (c) 2008, Martin W. Kirst All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#  Redistributions of source code must retain the above copyright notice,
#  this list of conditions and the following disclaimer.
#  Redistributions in binary form must reproduce the above copyright notice,
#  this list of conditions and the following disclaimer in the documentation
#  and/or other materials provided with the distribution.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
#  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
#  PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
#  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
#  TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
from __future__ import division, print_function

import errno
import socket

from ascii_telnet.ascii_movie import Movie

from ascii_telnet.ascii_player import VT100Player

try:
    # noinspection PyCompatibility
    from socketserver import StreamRequestHandler, ThreadingMixIn, TCPServer
except ImportError:  # Py2
    # noinspection PyCompatibility,PyUnresolvedReferences
    from SocketServer import StreamRequestHandler, ThreadingMixIn, TCPServer


class ThreadedTCPServer(ThreadingMixIn, TCPServer):
    daemon_threads = True


class TelnetRequestHandler(StreamRequestHandler):
    """
    Request handler used for multi threaded TCP server
    @see: SocketServer.StreamRequestHandler
    """

    filename = None  # filename is set once, so it's immutable and safe for multi threading

    def handle(self):
        movie = Movie()
        movie.load(TelnetRequestHandler.filename)

        self.player = VT100Player(movie)
        self.player.draw_frame = self.draw_frame
        self.player.play()

    def draw_frame(self, screen_buffer):
        """
        Gets the current screen buffer and writes it to the socket.
        """
        try:
            self.wfile.write(screen_buffer.read())
        except socket.error as e:
            if e.errno == errno.EPIPE:
                print("Client Disconnected.")
                self.player.stop()


ascii_theater/ascii_telnet_server.py
# coding=utf-8
# !/usr/bin/env python

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#  Copyright (c) 2008, Martin W. Kirst All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#  Redistributions of source code must retain the above copyright notice,
#  this list of conditions and the following disclaimer.
#  Redistributions in binary form must reproduce the above copyright notice,
#  this list of conditions and the following disclaimer in the documentation
#  and/or other materials provided with the distribution.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
#  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
#  PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
#  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
#  TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

"""
  ASCII art movie Telnet player.
  Version         : 0.1

  Can stream an ~20 minutes ASCII movie via Telnet emulation
  as stand alone server or via xinetd daemon.
  Tested with Python 2.6+, Python 3.5+

  Original art work : Simon Jansen ( http://www.asciimation.co.nz/ )
  Telnetification
  & Player coding   : Martin W. Kirst ( https://github.com/nitram509/ascii-telnet-server )
  Python3 Update: Ryan Jarvis

"""
from __future__ import division, print_function

import os
import sys
from optparse import OptionParser

from ascii_telnet.ascii_movie import Movie
from ascii_telnet.ascii_server import TelnetRequestHandler, ThreadedTCPServer

from ascii_telnet.ascii_player import VT100Player


def runTcpServer(interface, port, filename):
    """
    Start a TCP server that a client can connect to that streams the output of
     Ascii Player

    Args:
        interface (str):  bind to this interface
        port (int): bind to this port
        filename (str): file name of the ASCII movie
    """
    TelnetRequestHandler.filename = filename
    server = ThreadedTCPServer((interface, port), TelnetRequestHandler)
    server.serve_forever()


def runStdOut(filepath):
    """
    Stream the output of the Ascii Player to STDOUT
    Args:
        filepath (str): file path of the ASCII movie
    """

    def draw_frame_to_stdout(screen_buffer):
        sys.stdout.write(screen_buffer.read().decode('iso-8859-15'))

    movie = Movie()
    movie.load(filepath)
    player = VT100Player(movie)
    player.draw_frame = draw_frame_to_stdout
    player.play()


if __name__ == "__main__":
    usage = "Usage: %prog [options]"
    parser = OptionParser(usage=usage)
    parser.add_option("", "--standalone", dest="tcpserv", action="store_true",
                      help="Run as stand alone multi threaded TCP server (default)")
    parser.add_option("", "--stdout", dest="tcpserv", action="store_false",
                      help="Run with STDIN and STDOUT, for example in XINETD " +
                           "instead of stand alone TCP server. " +
                           "Use with python option '-u' for unbuffered " +
                           "STDIN STDOUT communication")
    parser.add_option("-f", "--file", dest="filename", metavar="FILE",
                      help="Text file containing the ASCII movie")
    parser.add_option("-i", "--interface", dest="interface",
                      help="Bind to this interface (default '0.0.0.0', all interfaces)",
                      default="0.0.0.0")
    parser.add_option("-p", "--port", dest="port", metavar="PORT",
                      help="Bind to this port (default 23, Telnet)",
                      default=23, type="int")
    parser.add_option("-v", "--verbose", action="store_true", dest="verbose",
                      help="Verbose (default for TCP server)")
    parser.add_option("-q", "--quiet", action="store_false", dest="verbose",
                      help="Quiet! (default for STDIN STDOUT server)")
    parser.set_defaults(interface="0.0.0.0",
                        port=23,
                        tcpserv=True,
                        verbose=True, )
    options = parser.parse_args()[0]

    if not (options.filename and os.path.exists(options.filename)):
        parser.exit(1, "Error, file not found! See --help for details.\n")

    try:
        if options.tcpserv:
            if options.verbose:
                print("Running TCP server on {0}:{1}".format(options.interface, options.port))
                print("Playing movie {0}".format(options.filename))
            runTcpServer(options.interface, options.port, options.filename)
        else:
            runStdOut(options.filename)
    except KeyboardInterrupt:
        print("Ascii Player Quit.")
